plat/mediatek/mt8183/drivers/gpio/mtgpio.c

```diff
@@ -0,0 +1,370 @@
+	else
+		mmio_write_32(DATAOUT_addr[pin].addr + 0x4, 1U << bit);
+}
+/*---------------------------------------------------------------------------*/
+int mt_get_gpio_out_chip(uint32_t pin)
+{
+	uint32_t bit;
+	uint32_t reg;
+
+	assert(pin < MAX_GPIO_PIN);
+
+	bit = pin % MAX_GPIO_REG_BITS;
+
+	/* To reduce memory usage, we don't use DATAOUT_addr[] array*/
+	reg = mmio_read_32(DATAOUT_addr[pin].addr);
+	return (((reg & (1L << bit)) != 0) ? 1 : 0);
+}
+/*---------------------------------------------------------------------------*/
+int mt_get_gpio_in_chip(uint32_t pin)
+{
+	uint32_t bit;
+	uint32_t reg;
+
+	assert(pin < MAX_GPIO_PIN);
+
+	bit = pin % MAX_GPIO_REG_BITS;
+
+	/* To reduce memory usage, we don't use DIN_addr[] array*/
+	reg = mmio_read_32(DIN_addr[pin].addr);
+	return (((reg & (1L << bit)) != 0) ? 1 : 0);
+}
+/*---------------------------------------------------------------------------*/
+void mt_set_gpio_mode_chip(uint32_t pin, int mode)
+{
+	uint32_t bit;
+	uint32_t data;
+	uint32_t mask;
+
+
+	assert(pin < MAX_GPIO_PIN);
+
+	assert(mode < GPIO_MODE_MAX);
+
+	mask = (1L << GPIO_MODE_BITS) - 1;
+	mode = mode & mask;
+	bit = (pin % MAX_GPIO_MODE_PER_REG) * GPIO_MODE_BITS;
+
+	data = mmio_read_32(MODE_addr[pin].addr);
+	data &= (~(mask << bit));
+	data |= (mode << bit);
+	mmio_write_32(MODE_addr[pin].addr, data);
+}
+/*---------------------------------------------------------------------------*/
+int mt_get_gpio_mode_chip(uint32_t pin)
+{
+	uint32_t bit;
+	uint32_t data;
+	uint32_t mask;
+
+	assert(pin < MAX_GPIO_PIN);
+
+	mask = (1L << GPIO_MODE_BITS) - 1;
+
+	bit = (pin % MAX_GPIO_MODE_PER_REG) * GPIO_MODE_BITS;
+	data = mmio_read_32(MODE_addr[pin].addr);
+	return (data >> bit) & mask;
+}
+/*---------------------------------------------------------------------------*/
+void mt_set_gpio_pull_enable_chip(uint32_t pin, int en)
+{
+	assert(pin < MAX_GPIO_PIN);
+
+	assert(!((PULLEN_offset[pin].offset == (int8_t)-1) && (PUPD_offset[pin].offset == (int8_t)-1)));
+
+	if (en == GPIO_PULL_DISABLE) {
+		if (PULLEN_offset[pin].offset == (int8_t)-1)
+			mmio_clrbits_32(PUPD_addr[pin].addr, 3L << PUPD_offset[pin].offset);
+		else
+			mmio_clrbits_32(PULLEN_addr[pin].addr, 1L << PULLEN_offset[pin].offset);
+	} else if (en == GPIO_PULL_ENABLE) {
+		if (PULLEN_offset[pin].offset == (int8_t)-1) {
+			/* For PUPD+R0+R1 Type, mt_set_gpio_pull_enable does not know
+			 * which one between PU and PD shall be enabled.
+			 * Use R0 to guarantee at one resistor is set when lk
+			 * apply default setting
+			 */
+			mmio_setbits_32(PUPD_addr[pin].addr, 1L << PUPD_offset[pin].offset);
+			mmio_clrbits_32(PUPD_addr[pin].addr, 1L << (PUPD_offset[pin].offset + 1));
+		} else {
+			/* For PULLEN + PULLSEL Type */
+			mmio_setbits_32(PULLEN_addr[pin].addr, 1L << PULLEN_offset[pin].offset);
+		}
+	} else if (en == GPIO_PULL_ENABLE_R0) {
+		assert(!(PUPD_offset[pin].offset == (int8_t)-1));
+		/* IOConfig does not support MWR operation, so use CLR + SET */
```

Comment:
I'm confused what this comment is trying to tell me. Can't you use mmio_clrsetbits_32?